### PR TITLE
Add support for look-around patterns using fancy-regex

### PR DESCRIPTION
**Pull request by duckontheweb**
Uses [`fancy-regex`](https://docs.rs/fancy-regex/0.5.0/fancy_regex/) in place of [`regex`](https://docs.rs/regex/) for matching against all patterns in JSON schemas. 

`regex` is still used when [converting a RegEx for ECMA 262](https://github.com/duckontheweb/jsonschema-rs/blob/b689ac5b32c29db0a36cf7ad3890e78f0cea96e3/jsonschema/src/keywords/pattern.rs#L78) to take advantage of the `Regex::replace_all` method, which is not included in `fancy-regex`.  Since `fancy-regex` depends on `regex` anyway, this does not affect the final size of the compiled binary (thanks to [this commit](https://github.com/niklasfasching/check_timed_logs_fast/commit/fa2ec6ef77d3befd03d8674e0ceae24846b620a6) for the idea). 

I also removed the test from #214, since that pattern is supported in this PR.

### CODE REVIEW

File: jsonschema/src/keywords/pattern.rs [hunk 1]
```diff
@@ -43,21 +41,25 @@ impl Validate for PatternValidator {
...
     }
 
     fn is_valid(&self, _: &JSONSchema, instance: &Value) -> bool {
         if let Value::String(item) = instance {
-            if !self.pattern.is_match(item) {
-                return false;
+            if let Ok(is_match) = self.pattern.is_match(item) {
+                if !is_match {
+                    return false;
+                }
```
Comment (Stranger6667) [inline jsonschema/src/keywords/pattern.rs:62]: I think that it could be 

```suggestion
                return is_match
```

File: jsonschema/src/keywords/pattern.rs [hunk 2]
```diff
@@ -43,21 +41,25 @@ impl Validate for PatternValidator {
...
             }
         }
         no_error()
     }
 
     fn is_valid(&self, _: &JSONSchema, instance: &Value) -> bool {
         if let Value::String(item) = instance {
-            if !self.pattern.is_match(item) {
-                return false;
+            if let Ok(is_match) = self.pattern.is_match(item) {
```
Comment (Stranger6667) [inline jsonschema/src/keywords/pattern.rs:59]: What about the `Err` variant? I believe it should lead to `false`.

File: jsonschema/src/keywords/pattern.rs [hunk 3]
```diff
@@ -43,21 +41,25 @@ impl Validate for PatternValidator {
         instance_path: &InstancePath,
     ) -> ErrorIterator<'a> {
         if let Value::String(item) = instance {
-            if !self.pattern.is_match(item) {
-                return error(ValidationError::pattern(
-                    instance_path.into(),
-                    instance,
-                    self.original.clone(),
-                ));
+            if let Ok(is_match) = self.pattern.is_match(item) {
```
Comment (Stranger6667) [inline jsonschema/src/keywords/pattern.rs:44]: If `Err` is returned, we should indicate this backtracking problem to the user with a new `ValidationError` variant.

**Review by Stranger6667 — CHANGES_REQUESTED** [review]
I like the direction and it looks good to me :)

I think that it is important to add a new variant to the `ValidationErrorKind` enum that will indicate the backtracking issue and use it in `PatternValidator::validate` and other cases when patterns are used (`additionalProperties`)

**Comment by Stranger6667** [comment]
Maybe one extra thing - it will be nice to have a test case for `pattern` with backtracking

File: jsonschema/src/keywords/pattern.rs (hunk 1 above)
Comment (duckontheweb) [reply jsonschema/src/keywords/pattern.rs:62]: Yes, thanks!

File: jsonschema/src/keywords/pattern.rs (hunk 3 above)
Comment (duckontheweb) [reply jsonschema/src/keywords/pattern.rs:44]: Yes, good idea. I'll add this.

File: jsonschema/src/keywords/pattern.rs (hunk 2 above)
Comment (duckontheweb) [reply jsonschema/src/keywords/pattern.rs:59]: Yeah, I had a hard time reasoning out what would be expected here. I started typing out a response here and then came to same conclusion that it should lead to `false`. I'll update that.

**Comment by duckontheweb** [comment]
@Stranger6667 How would you like me to handle the commits to address these comments. Should I squash them with the existing commit?

**Comment by Stranger6667** [comment]
In the end, I plan to squash them into one, so feel free to push more :)

**Comment by duckontheweb** [comment]
I think I addressed all of the suggestions above via the new commits. I also added matching and non-matching test cases to `pattern.rs`.

I had a thought that it would be good to add an integration test for the combination of `patternProperties`, `properties`, and `additionalProperties` that I mentioned in [this comment](https://github.com/Stranger6667/jsonschema-rs/issues/183#issuecomment-831473463), but I wasn't sure how to fit it in with the official JSON Schema test suite.

File: jsonschema/src/keywords/pattern.rs [hunk 4]
```diff
@@ -43,35 +41,44 @@ impl Validate for PatternValidator {
...
+            return self.pattern.is_match(item).unwrap_or(false);
         }
         true
     }
 }
 
 impl ToString for PatternValidator {
     fn to_string(&self) -> String {
-        format!("pattern: {}", self.pattern)
+        format!("pattern: {:?}", self.pattern)
```
Comment (Stranger6667) [inline jsonschema/src/keywords/pattern.rs:76]: Hm, is it required?

File: jsonschema/src/keywords/pattern.rs [hunk 5]
```diff
@@ -110,14 +117,16 @@ fn convert_regex(pattern: &str) -> Result<Regex, regex::Error> {
...
 
 #[allow(clippy::integer_arithmetic)]
-fn replace_control_group(captures: &Captures) -> String {
+fn replace_control_group(captures: &regex::Captures) -> String {
     // There will be no overflow, because the minimum value is 65 (char 'A')
     ((captures
-        .index(0)
+        .get(0)
+        .map(|m| m.as_str())
+        .expect("index 0 to return the whole match")
```
Comment (Stranger6667) [inline jsonschema/src/keywords/pattern.rs:129]: As the type didn't change, why are these changes needed?

File: jsonschema/src/keywords/pattern_properties.rs [hunk 6]
```diff
@@ -80,7 +80,9 @@ impl ToString for PatternPropertiesValidator {
             "patternProperties: {{{}}}",
             self.patterns
                 .iter()
-                .map(|(key, validators)| { format!("{}: {}", key, format_validators(validators)) })
+                .map(|(key, validators)| {
+                    format!("{:?}: {}", key, format_validators(validators))
```
Comment (Stranger6667) [inline jsonschema/src/keywords/pattern_properties.rs:84]: Is the format change needed?

File: jsonschema/src/keywords/pattern_properties.rs [hunk 7]
```diff
@@ -148,7 +150,7 @@ impl Validate for SingleValuePatternPropertiesValidator {
 impl ToString for SingleValuePatternPropertiesValidator {
     fn to_string(&self) -> String {
         format!(
-            "patternProperties: {{{}: {}}}",
+            "patternProperties: {{{:?}: {}}}",
```
Comment (Stranger6667) [inline jsonschema/src/keywords/pattern_properties.rs:153]: Is the format change needed?

File: jsonschema/src/keywords/pattern.rs (hunk 4 above)
Comment (duckontheweb) [reply jsonschema/src/keywords/pattern.rs:76]: These were required because `fancy_regex::Regex` implements `Debug` but not `Display`.

Comment (Stranger6667) [reply jsonschema/src/keywords/pattern.rs:76]: Aha, got it :) then it obsoletes some of my other comments

File: jsonschema/src/keywords/pattern.rs (hunk 5 above)
Comment (duckontheweb) [reply jsonschema/src/keywords/pattern.rs:129]: Good catch. This got left over from my initial attempt to use `fancy_regex::Captures` here. Now that we're continuing to use `regex` this is no longer necessary, so I'll revert it.

File: jsonschema/src/keywords/pattern_properties.rs (hunk 7 above)
Comment (duckontheweb) [reply jsonschema/src/keywords/pattern_properties.rs:153]: Same as above, this is required because `fancy_regex::Regex` implements `Debug` but not `Display`.

**Review by Stranger6667 — APPROVED** [review]

**Comment by Stranger6667** [comment]
Looks great to me! :) Thank you for contributing. I appreciate it :)

I'll check it one more time tomorrow and probably will release these changes in the new release :)

**Comment by duckontheweb** [comment]
I'm not sure what to make of that `clippy` failure. It doesn't show up when I run `cargo clippy --all-targets --all-features -- -D warnings` locally, but I'm on a Mac. I'll try running in an Ubuntu container as well to see if I can replicate it.

**Comment by duckontheweb** [comment]
> Looks great to me! :) Thank you for contributing. I appreciate it :)
> 
> I'll check it one more time tomorrow and probably will release these changes in the new release :)

No problem, thanks for bearing with me on this. It's my first contribution to a Rust project and it's been really educational to see how this project is set up.

**Comment by Stranger6667** [comment]
> I'm not sure what to make of that clippy failure. It doesn't show up when I run cargo clippy --all-targets --all-features -- -D warnings locally, but I'm on a Mac. I'll try running in an Ubuntu container as well to see if I can replicate it.

Hm, maybe it is connected to the recent Rust toolchain update :shrug: I'll apply a fix after merge :)

```
pub(crate) const fn backtrack_limit(
```

**Comment by Stranger6667** [comment]
> No problem, thanks for bearing with me on this. It's my first contribution to a Rust project and it's been really educational to see how this project is set up.

I am glad to hear that! :)
